### ISSUE TITLE
[Bug]: Chain Permissions - When switching networks, either through the wallet interface or directly from a dApp, the change does not apply globally across all instances

### ISSUE THREAD

**Issue by sleepytanya** [MetaMask/metamask-mobile#12301]
### Describe the bug

When switching networks, either through the wallet interface or directly from a dApp, the change does not apply globally across all instances.

[Permissions APK build](https://drive.google.com/file/d/1gvTwY06OH9OnWXGsg6w6q9tKc33DPHbf/view)

1. Selecting network in the wallet - supposed to change networks for all dapps.
       In current Permissions build it doesn't work as expected. The MM icon displays network selected in the wallet but the actual network for the dapp is previously selected chain. Tx created on the previous network and fails with the 'current chain doesn't match' error.

https://github.com/user-attachments/assets/a21c1550-829a-426b-a4ee-a84f7474d91f

![Screenshot 2024-11-12 at 22 56 42](https://github.com/user-attachments/assets/82e43880-3558-46b0-b297-5debbc56cc33)

![Screenshot 2024-11-13 at 00 15 49](https://github.com/user-attachments/assets/66ec97cd-834b-40d1-a1b5-8ab2415877f9)

2. Changing network from the dapp - also supposed to change selected network globally. Looks like it doesn't work as well.

Steps in screen recording:

- all dapps on Base initially
- I switch from Sushiswap to Polygon
       Result:
- wallet is on Polygon
- all dapps in the menu are on Polygon
- test dapp displays Poly icon but chain id0x2105 (Base), tx can't be created, all buttons are non-responsive
- Pancake swap displays Poly icon but chain is Base (tx created on Base)

https://github.com/user-attachments/assets/76a0958a-b3f6-4b31-9be8-0b641537f9f0

![Screenshot 2024-11-13 at 00 05 28](https://github.com/user-attachments/assets/dc14c810-7462-41c4-934d-62f263ec28e4)

### Expected behavior

-

### Steps to reproduce

1. See the description

### Detection stage

On the development branch

### Version

[Permissions APK build](https://drive.google.com/file/d/1gvTwY06OH9OnWXGsg6w6q9tKc33DPHbf/view)

### Build type

None

### Device

Samsung S24+

### Operating system

Android

**Comment by sleepytanya** [comment]
Attempt to reload the dapp page manually doesn't resolve the issue - the chain logo updates to the correct one but transactions are still created on the previously selected network:


https://github.com/user-attachments/assets/0b7fc227-f26b-414e-a1be-65e9aab16f3a

**Comment by sleepytanya** [comment]
@EtherWizard33 

> found the cause
TLDR; not related to my work
Someone made a change 3 weeks ago to the webview and broke it 
https://github.com/MetaMask/metamask-mobile/pull/11976

**Comment by sleepytanya** [comment]
7.24.2, June 15, 2024 - network switching works normally.
Switching network from the wallet and from the dapps applies selected network globally.  Icons/chain ids match, tx created on the globally selected network


https://github.com/user-attachments/assets/61623bab-9bcb-43a2-aa38-6a40186d6cdd


[Bitrise Release build](https://app.bitrise.io/app/be69d4368ee7e86d/installable-artifacts/8ea190215436d184),  latest `main `
I see the problem with switching networks but only see it with Pancakeswaps and seems it only happens when I switch to certain network/-s. 

Steps to reproduce:
-connect to Pancakeswaps
- switch to Ethereum / Arbitrum / smthg else, observe that network successfully switched globally
- go back to Pancakeswaps and switch to BNB / Base / etc 
- observe that network switched in the wallet, in the other dapps, but Pancakeswaps network symbol stuck on loading spinner
- issue can be resolved by manually reloading the page



https://github.com/user-attachments/assets/39fd5394-9ba9-427e-8933-4e979784acf4



https://github.com/user-attachments/assets/36f05426-f4e7-4178-9cd5-9ff15f79b6ac

**Comment by EtherWizard33** [comment]
@sleepytanya the network switch issue was specific to that build, the cause was the MM_INFURA_PROJECT_ID was not set.
This issue should not be seend in production. If you are ok with the new build i will create, this issue will then be closable since it should have been specific to the previous QA build.

**Comment by sleepytanya** [comment]
@EtherWizard33 

Posting update here since it's too long to slack.

There appears to be a persistent issue related to page reloading / network switching that does not fully align with expectations. 

1. iOS
 - wallet + dapps on Base
 - from the pancake swap I switch to BNB 
 - Portfolio, pancake + wallet are on BNB
 - sushiswap page shows correct chain icon but the dapp is still connected to Base (you can see the incorrect Eth balance)
 - Reload doesn't resolve the issue
 -  I have to manually switch sushiswap dapp to BNB
 


https://github.com/user-attachments/assets/fbca6c52-0370-46a6-b932-4bd7a61f7b87


2. Android
- Wallet + dapps on BNB
- from pancakeswap I switch to Linea
- wallet, pancakeswap, Portfolio are on Linea
- sushiswap page behaves the same as on iOS: it stays on the previous network until I manually switch to Linea



https://github.com/user-attachments/assets/7cc5f9b6-f214-4692-a0b4-c034a9ae1021


3. Android
- wallet + dapps are on Linea
- from the sushiswap page I switch to Base
- sushi page shows correct Base balance, correct Base wallet icon but incorrect dapp chain icon (it's still LInea)
- wallet + portfolio + pancake are on Base
- from the pancake page I switch to BNB
- pancake shows correct BNB balance, correct wallet icon but page chain icon stays loading until I reload page
- sushiswap behavior is the same as in above cases: the page stays connected to the previous network until I manually switch networks



https://github.com/user-attachments/assets/2d9e5c6b-4a37-4cbe-9112-2b603db5178f




<img width="329" alt="Screenshot 2024-11-28 at 00 49 13" src="https://github.com/user-attachments/assets/3371a479-b131-45e4-b60f-495667b78400">


<img width="339" alt="Screenshot 2024-11-28 at 00 49 34" src="https://github.com/user-attachments/assets/c22b7d60-4d61-405d-bc57-5dae87636726">

**Comment by hesterbruikman** [comment]
1 & 2 seem like a Sushi swap issue. The network indicator in Sushiswap shows BNB, the main button shows `Switch to Base` or BNB chain respectively. This gives two indicators that the dapp is aware of the wallet active network respectively being BNB or Linea. The asset values not showing correctly seems more likely an issue on the dapp side.

Generally, these seem cases where dapps don't update upon changes in the wallet. It's a standards and interfacing issue and we can't guarantee that a dapp will update:

- Not all dapps respond to `chainChanged` or `accountChanged` callbacks as not all wallets return this information. There might be a pattern in how libraries handle this (cc @vandan @adamceresko)
- dapps might have different triggers to reload, e.g. they might only update when tapping an action in their UI like Swap or Supply. From what I understand, best we can do is ensure manual reload is always available and easily discoverable (this we can improve) and perhaps cache active chain and account

The upside is that with full multichain and chain permissions, ultimately there will not a be concept of an `active chain` and these issues will become less visible. Uniswap and Squidrouter are examples of where you already see this concept abstracted away.

**Comment by hesterbruikman** [comment]
Downgrading this to `sev2`. We're aware of the suboptimal mirroring of network between wallet and dapp. This is not fully within our control as it depends on dapp server side reload and response to chainChanged. In progress work to store a 'per dapp selected network' should for the most part resolve this issue in the coming months.
